src/mpi/comm/comm_impl.c

```diff
@@ -795,6 +795,40 @@ int MPIR_Comm_create_from_group_impl(MPIR_Group * group_ptr, const char *stringt
     goto fn_exit;
 }
 
+/* a restricted implementation of MPI_Intercomm_create_from_groups.
+ * Require comm_world, and remote_group part of comm_world.
+ * TODO: remote_group from different comm_world
+ */
+int MPIR_Intercomm_create_from_groups_impl(MPIR_Group * local_group_ptr, int local_leader,
+                                           MPIR_Group * remote_group_ptr, int remote_leader,
+                                           const char *stringtag,
+                                           MPIR_Info * info_ptr, MPIR_Errhandler * errhan_ptr,
+                                           MPIR_Comm ** p_newintercom_ptr)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_Assert(MPIR_Process.comm_world);
+
+    MPIR_Comm *local_comm;
+    mpi_errno = MPIR_Comm_create_from_group_impl(local_group_ptr, stringtag, info_ptr, errhan_ptr,
+                                                 &local_comm);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    int tag = get_tag_from_stringtag(stringtag);
+    int remote_lpid = remote_group_ptr->lrank_to_lpid[remote_leader].lpid;
+    mpi_errno = MPIR_Intercomm_create_impl(local_comm, local_leader,
```

Comment:
@hzhou In here, you are assuming that `remote_grup` is a subset of of the group of `MPIR_Process.comm_world`. There may be details that I'm missing about how MPICH implement things, but if `MPIR_Process.comm_world` is just an internal representation of the old good `MPI_COMM_WORLD`, then I believe this implementation is wrong.

Imagine I use dynamic process management to spawn new processes. Then I merge the intercommunicator to get an intracommunicator with both parents and child processes. Form that big intracomm, I get the group and use it to call this new MPI-4 `MPI_Intercomm_create_from_groups` routine passing it as `remote_group`. Then `remote_group` is not a subset of `comm_world`. Would your current implementation still work in this case?

Reply:
Yeah, it won't work for now. We need to overhaul the current dynamic process management in order to implement this properly and cleanly. 

On the other hand, I don't see it necessary that a user application have to use `MPIR_Intercomm_create_from_groups`, vs. just use `MPIR_Intercomm_create`. So I think it is okay to postpone the full implementation for a bit.